libraries/Schedule/src/circular_queue/circular_queue.h

```diff
@@ -0,0 +1,438 @@
+/*
+circular_queue.h - Implementation of a lock-free circular queue for EspSoftwareSerial.
+Copyright (c) 2019 Dirk O. Kaar. All rights reserved.
+
+This library is free software; you can redistribute it and/or
+modify it under the terms of the GNU Lesser General Public
+License as published by the Free Software Foundation; either
+version 2.1 of the License, or (at your option) any later version.
+
+This library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public
+License along with this library; if not, write to the Free Software
+Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#ifndef __circular_queue_h
+#define __circular_queue_h
+
+#include <atomic>
+#include <memory>
+#include <algorithm>
+#include <functional>
+#ifdef ESP8266
+#include "interrupts.h"
+#else
+#include <mutex>
+#endif
+
+#ifdef ESP32
+#include <esp_attr.h>
+#elif !defined(ESP8266)
+#define ICACHE_RAM_ATTR
+#define IRAM_ATTR
+#endif
+
+/*!
+    @brief	Instance class for a single-producer, single-consumer circular queue / ring buffer (FIFO).
+            This implementation is lock-free between producer and consumer for the available(), peek(),
+            pop(), and push() type functions.
+*/
+template< typename T > class circular_queue
```

Comment:
This file is very difficult to understand. I was trying to look at the class interface, and there is no way to tell at a glance how it's meant to be used. 
I suspect the monolithic nature of the file is convenience for porting, but here we need maintainability. 
I suggest:
1. splitting the single consumer and multiconsumer classes into their own specific files
2. Splitting the template class method implementations into separate .h (usually blah_impl.h naming convention) 
3. including the above headers into this circular_queue.h header

Reply:
Also, please put the template <> test on its own line rather than precursor to class blah

Reply:
Thank you so much for your input, splitting up the two classes into their own source files is rather obvious, and I forgot doing it. It's a clear maintenance improvement.
For the other requests, I have a strong opinion on those. The template<> prefix occurs once in each file, and makes it rather more expressive in my reading - I would not want to change that, given that the resulting line lengths are not an issue, either.
Second, I am generally opposed to creating a plethora of single source files, which I find extremely hard to deal with in any editor, when there is no overriding reason to do so. While a separate interface would definitely be helpful, I see so reasonably easy way of doing so in non-virtual C++ classes - for lack of the interface keyword. If there's a pattern to do so that I missed, please help me out. Introducing the usual pure-virtual abstract base class in code for an MCU in my mind is out of the question.
So, while removing the inline-definitions of the member function into another source file (*_impl.h), I would find code editing and comprehension a pain. YMMV (respestively does vary), no hard feeling, but could you find a way to agree to keeping this to my preference?

At the end of the day, I was hoping to have the "Schedule" code an own library that can be plugged, you've seen #6182 prepare for that.

Reply:
What are you talking about with virtualness, there is no virtualness here.

The current circular_queue.h is 281 lines. Splitting it will increase it and probably bump it over 300. As a rule of thumb, it's better to have multiple smaller files than one big one. 
For a normal class, the usual implementation is to put the class declaration in the .h and the method definitions in the .cpp. Of course this isn't possible for a template class, so the next best thing is to put the template class declaration in one .h, and the template method definitions in another *_impl.h which then gets #included at the end of the definition one.
Despite the length, I can live with one .h for both declaration and definition here, but the minimum is to split the class declaration from the definition to make the class interface readable at all.
About template and class on different lines, we (maintainers) have put forth large effort to clean up style and layout throughout the core, and maintain consistency thereafter. The request is merely to change to this:
```cpp
template <typename arg>
class circular_queue
{
...
};
```
Please comply with that.

Reply:
I agree, so I have moved the definitions of each member function that has more than 4-5 lines of code outside the class definition. I hope you like this, too.